adr/ADR-15.md

```diff
@@ -0,0 +1,137 @@
+	RateLimit       uint64        `json:"rate_limit_bps,omitempty"` // Bits per sec
+	SampleFrequency string        `json:"sample_freq,omitempty"`
+	MaxWaiting      int           `json:"max_waiting,omitempty"`
+	MaxAckPending   int           `json:"max_ack_pending,omitempty"`
+	FlowControl     bool          `json:"flow_control,omitempty"`
+	Heartbeat       time.Duration `json:"idle_heartbeat,omitempty"`
+}
+```
+
+#### Stream name
+
+When no stream name is specified, the library will use the subject provided as a way to find out which stream this subscription is for. A request is sent to the server on the `<JS prefix>.STREAM.NAMES` subject with a JSON content `{"subject":"<subject here>"}`. If the response (a list of stream names) is positive and contains a single entry, then the library will use this stream name, otherwise an error indicating that there is no matching stream name should be returned.
+
+#### Consumer name
+
+When no consumer name is specified, but a configuration is provided, the consumer will be the `Durable` field of the `ConsumerConfig` object.
+
+Note if the user provides a `Durable` in the configuration, this is treated differently that if the user provides a consumer to the subscribe API. Providing the stream and consumer name through the API makes the subscription being "bound", that is, no attempt to create a JS consumer will be done because the user expects the stream and consumer to already exists.
+
+So if the consumer is not provided but set to the `ConsumerConfig.Durable` field, this means that if the durable exists, it will be attached to it, but if it doesn't, then the library will try to create this durable.
+
+After this step, the library has a stream name and a consumer name, the latter possibly empty. If a consumer name is provided, the library will lookup the consumer info based on the pair (stream, consumer). If a consumer info is returned, some checks should be made so that the user provided `ConsumerConfig` matches what is returned by the server. Namely:
+
+- The `FilterSubject` matches the subject passed to the "Subscribe" API, if not return an error indicating the subject mismatch
+- If the user was trying to create a Pull subscription, but `DeliverSubject` is not empty, then return an error indicating that user can't create a pull subscription for a push based JS consumer
+- The opposite: if the user was not creating a Pull subscription but the `DeliverSubject` is empty, then return an error indicating that a pull susbcription is required
+
+If no error is found, then the internal NATS subscription should be created on the returned `DeliverSubject`.
+
+If there was a consumer lookup but an error occurs, the behavior depends on few things:
+
+- If the error is a "not found" (that is, the library got a response indicating that the said consumer does not exists):
+    - If the subscription is "bound" (stream and consumer names were provided), then unless this is a "pull" subscription, the error is returned.
```

Comment:
bind is valid for both pull and push. If bind is requested and the consumer for the stream/durable is not found, it is an error.